pkg/queue/certificate/watcher.go

```diff
@@ -0,0 +1,111 @@
+package certificate
+
+import (
+	"crypto/tls"
+	"fmt"
+	"path"
+	"sync"
+
+	"github.com/fsnotify/fsnotify"
+	"go.uber.org/zap"
+)
+
+// CertWatcher watches certificate and key files and reloads them if they change on disk.
+type CertWatcher struct {
+	logger *zap.SugaredLogger
+
+	certPath    string
+	keyPath     string
+	certificate *tls.Certificate
+
+	stop    chan struct{}
+	watcher *fsnotify.Watcher
+	mux     sync.RWMutex
+}
+
+// NewCertWatcher creates a CertWatcher
+func NewCertWatcher(certPath, keyPath string, logger *zap.SugaredLogger) (*CertWatcher, error) {
+	cw := &CertWatcher{
+		logger:   logger,
+		certPath: certPath,
+		keyPath:  keyPath,
+		stop:     make(chan struct{}),
+		mux:      sync.RWMutex{},
+	}
+
+	certDir := path.Dir(cw.certPath)
+	keyDir := path.Dir(cw.keyPath)
+
+	cw.logger.Info("Starting to watch the following directories for changes",
+		zap.String("certDir", certDir), zap.String("keyDir", keyDir))
+
+	var err error
+
+	if cw.watcher, err = fsnotify.NewWatcher(); err != nil {
+		return nil, fmt.Errorf("failed to create a file watcher: %w", err)
+	}
+
+	if err = cw.watcher.Add(certDir); err != nil {
+		return nil, fmt.Errorf("failed to add watch for %s: %w", certDir, err)
+	}
+
+	if keyDir != certDir {
+		if err = cw.watcher.Add(keyDir); err != nil {
+			return nil, fmt.Errorf("failed to add watch for %s: %w", keyDir, err)
+		}
+	}
+
+	cw.loadCert()
+
+	go cw.watch()
+
+	return cw, nil
+}
+
+func (cw *CertWatcher) watch() {
+	for {
+		select {
+
+		case <-cw.stop:
+			return
+
+		case _, ok := <-cw.watcher.Events:
+			if !ok {
+				// only happens when watcher is stopped
+				return
+			}
+			cw.loadCert()
+
+		case err, ok := <-cw.watcher.Errors:
+			if !ok {
+				// only happens when watcher is stopped
+				return
+			}
+			cw.logger.Error("received an error from file watcher", zap.Error(err))
+		}
+	}
+}
+
+// Stop shuts down the CertWatcher. Use this with `defer`.
+func (cw *CertWatcher) Stop() {
+	cw.logger.Info("Stopping file watcher")
+	close(cw.stop)
+	defer cw.watcher.Close()
+}
+
+func (cw *CertWatcher) loadCert() {
+	keyPair, err := tls.LoadX509KeyPair(cw.certPath, cw.keyPath)
+	if err != nil {
+		cw.logger.Error("failed to load certificate", zap.Error(err))
+	} else {
+		cw.mux.Lock()
+		defer cw.mux.Unlock()
+		cw.certificate = &keyPair
+		cw.logger.Info("Certificate and key updated")
+	}
+}
+
+// GetCertificate returns the server certificate for a client-hello request.
+func (cw *CertWatcher) GetCertificate(_ *tls.ClientHelloInfo) (*tls.Certificate, error) {
+	return cw.certificate, nil
```

Comment:
@ReToCode shouldn't you use the mutex here, too ? Or otherwise, why have a mutex then at all, since `loadCert` can only be called sequentially (only from the `watch` subroutine) anyway ?